Gradle: Fix dependency resolution

The Gradle plugin needs to gather all jars in order to perform a classpath scan and analyze the npm-related java annotations.

Currently, this is done in the `vaadinPrepareFrontend` task: a set of dependencies is figured out, from that a set of projects, from that a set of `assemble` tasks producing those jars. However, this does not work reliably:

1. Doesn't work for transitive `implementation` dependencies
2. Doesn't work if the jar file was not produced by `assemble` task.

Solution is to simplify the code and take advantage of the fact that Gradle can depend on a FileCollection. The line in the `VaadinPrepareFrontendTask` class will then simplify to

`dependsOn(project.configurations.runtimeClasspath.jars)`

```kotlin
/**
 * Returns the "runtimeClasspath" file collection.
 */
internal val ConfigurationContainer.runtimeClasspath: Configuration
    get() = getByName("runtimeClasspath")

/**
 * Returns only jar files from given file collection.
 */
internal val Configuration.jars: FileCollection
    get() = filter { it.name.endsWith(".jar", true) }
```

See this commit for more details: https://github.com/vaadin/vaadin-gradle-plugin/commit/62211dfe80ea824a782d46ef3da1c1c7aabf37f5

### (Possible) steps to reproduce: ###

Have a Gradle multiproject which contains WAR project which implementation-depends on another project which implementation-depends on another project. At the moment the plugin will not not depend on the third project but will still try to read the outcome jar of that project, leading to the following exception:

```
Caused by: java.lang.IllegalStateException: xyz.jar doesn't exist
        at com.vaadin.gradle.GradlePluginAdapter.getClassFinder(GradlePluginAdapter.kt:68)
        at com.vaadin.flow.plugin.base.BuildFrontendUtil.prepareFrontend(BuildFrontendUtil.java:137)
        at com.vaadin.gradle.VaadinPrepareFrontendTask.vaadinPrepareFrontend(VaadinPrepareFrontendTask.kt:71)
```

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/GradlePluginAdapter.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/GradlePluginAdapter.kt
@@ -83,10 +83,7 @@ internal class GradlePluginAdapter(val project: Project, private val isBeforePro
     }
 
     override fun getJarFiles(): MutableSet<File> {
-        val jarFiles = project.configurations
-            .getByName("runtimeClasspath")
-            .resolve()
-            .filter { it.name.endsWith(".jar", true) }
+        val jarFiles: Set<File> = project.configurations.runtimeClasspath.jars.toSet()
         return jarFiles.toMutableSet()
     }
 

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt
@@ -43,13 +43,7 @@ public open class VaadinPrepareFrontendTask : DefaultTask() {
         // the Vaadin classpath scanning will not work properly. See
         // https://github.com/vaadin/vaadin-gradle-plugin/issues/38
         // for more details.
-        val dependentProjectTasks: List<Task> = project.configurations.getByName("runtimeClasspath")
-                .allDependencies
-                .withType(ProjectDependency::class.java)
-                .toList()
-                .map { it.dependencyProject }
-                .mapNotNull { it.tasks.findByName("assemble") }
-        dependsOn(*dependentProjectTasks.toTypedArray())
+        dependsOn(project.configurations.runtimeClasspath.jars)
     }
 
     @TaskAction

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinUtils.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinUtils.kt
@@ -15,6 +15,10 @@
  */
 package com.vaadin.gradle
 
+
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.ConfigurationContainer
+import org.gradle.api.file.FileCollection
 import com.vaadin.flow.server.frontend.FrontendTools
 import com.vaadin.flow.server.frontend.FrontendUtils
 import org.gradle.api.Project
@@ -68,3 +72,15 @@ internal fun VaadinFlowPluginExtension.createFrontendTools(): FrontendTools =
         nodeVersion,
         URI(nodeDownloadRoot)
     )
+
+/**
+ * Returns the "runtimeClasspath" file collection.
+ */
+internal val ConfigurationContainer.runtimeClasspath: Configuration
+    get() = getByName("runtimeClasspath")
+
+/**
+ * Returns only jar files from given file collection.
+ */
+internal val Configuration.jars: FileCollection
+    get() = filter { it.name.endsWith(".jar", true) }

